File: src/csharp/eu/sig/training/ch02/LevelHelper.cs
namespace eu.sig.training.ch02
{

    public class LevelHelper
    {
        private bool inProgress;

        // tag::firstStepStart[]
        public void Start()
        {
            if (inProgress)
            {
                return;
            }
            inProgress = true;
        }
        // end::firstStepStart[]
    }

    public class ExtractMethod
    {
        private bool inProgress;

        // tag::extractMethodStart[]
        public void Start()
        {
            if (inProgress)
            {
                return;
            }
            inProgress = true;
            UpdateObservers();
        }
        // end::extractMethodStart[]

        // tag::updateObservers[]
        public void UpdateObservers()
        {
            UpdateObserversPlayerDied();
            UpdateObserversPelletsEaten();
        }
        // end::updateObservers[]

        private void UpdateObserversPlayerDied() { }

        private void UpdateObserversPelletsEaten() { }

    }

}

File: src/java/eu/sig/training/ch02/LevelHelper.java
package eu.sig.training.ch02;

public class LevelHelper {
    private boolean inProgress;

    // tag::firstStepStart[]
    public void start() {
        if (inProgress) {
            return;
        }
        inProgress = true;
    }
    // end::firstStepStart[]
}

class ExtractMethod {
    private boolean inProgress;

    // tag::extractMethodStart[]
    public void start() {
        if (inProgress) {
            return;
        }
        inProgress = true;
        updateObservers();
    }
    // end::extractMethodStart[]

    // tag::updateObservers[]
    private void updateObservers() {
        updateObserversPlayerDied();
        updateObserversPelletsEaten();
    }
    // end::updateObservers[]

    private void updateObserversPlayerDied() {}

    private void updateObserversPelletsEaten() {}
}
